Finish the sec_key -> signature work

danbernier opened the pull request:
This also includes all the commits from https://github.com/smileidentity/smile-identity-core-ruby/pull/13 Once that's merged into master and back into the `signature` branch, I'll probably rebase this, and there'll be a lot less noise. If you want to review before then, the commits are all ordered: refactors first, sec_key/signature work after...34449fd is the first signature-related commit here.

danbernier committed:
Signature: group sec_key and signature methods together

danbernier committed:
Signature#confirm_signature: reuse generation method

danbernier committed:
Refactor WebApi#options= to streamline

The paragraph for a nil options can be collapsed into a `||`.

danbernier committed:
Utilities job-status: collect options-munging into one place

Tightened up the specs around it, too. NB: the specs pointed out that we were
defaulting options[:return_job_status] to false, but never actually reading its
value. The README talks about it in the context of WebApi, so I'm pretty sure
this was a copy-paste error.

danbernier committed:
Delete empty spec file

danbernier committed:
Utilities spec: delete unused 'let(:timestamp)'

You can verify by replacing it with a `raise` in its body. It's never called.

danbernier committed:
Utilities: add option to :use_legacy_sec_key, or not

danbernier committed:
Specs: remove unneeded metaprogramming for initialization

danbernier committed:
Utilities spec: fix description: query_job_status ignores job_complete

danbernier committed:
Utilities spec: ensure we raise when sec_key is invalid

danbernier committed:
IdApi spec, symbolize_keys: empty describe -> pending spec

danbernier committed:
IdApi, WebApi specs: fix whitespace: long lines, empty lines, etc

danbernier committed:
IdApi, WebApi specs: use {}.merge instead of clone

danbernier committed:
IdApi configure_json spec: rewrite to be more direct

danbernier committed:
IdApi: add option to :use_legacy_sec_key, or not

danbernier committed:
WebApi specs: no-op the sleep for a faster spec run

danbernier committed:
WebApi: rename isImageFile? -> image_file?, fix spacing

danbernier committed:
WebApi: add option to :use_legacy_sec_key, or not

danbernier committed:
WebApi specs: streamline mocks for setup_requests

A lot of the setup was obscuring what we're actually testing.

danbernier committed:
WebApi specs for configure_info_json: streamline and tighten

danbernier committed:
Specs: update timestamp regexes for +0000 timezones

beaglebets commented on lib/smile-identity-core/id_api.rb:
```diff
@@ -17,12 +17,10 @@ def initialize(partner_id, api_key, sid_server)
-      @timestamp = Time.now.to_i
-
       self.id_info = symbolize_keys id_info
+      @use_legacy_sec_key = symbolize_keys(options || {}).fetch(:use_legacy_sec_key, true)
```
why would you need `options || {}` if your are assigning options as an empty hash when not provided on line 20?

beaglebets replied:
the option is `signature`. We unfortunately can't default to it, because then it breaks people's callbacks or existing implementations because `sec_key` isn't there. Why is the option called `signature`? because that's what the initial implementation was and we want the libraries to all look the same, and have things named identically.

danbernier replied:
Okay, that makes sense. I should've checked the other library work first. FWIW, it does currently default to using the sec_key, as we discussed.

danbernier replied:
> why would you need options || {} if your are assigning options as an empty hash when not provided on line 20?

Because it can still be passed as nil; setting it in the signature only determines what happens when it's not provided.

danbernier replied:
@beaglebets I changed `use_legacy_sec_key` to `signature` in 3fcdff8.

beaglebets commented on lib/smile-identity-core/id_api.rb:
```diff
@@ -98,21 +96,29 @@ def setup_requests
-
-      body.merge!(@id_info)
-      JSON.generate(body)
+    def id_verification_request
```
I think we should revert the function names. The function names in this library match the function names in all the other libraries. This is by design, so when someone is looking to implement the same functionality in a different library, we are all using the same verbiage.

beaglebets replied:
This comment applies throughout.

danbernier replied:
Changed back in f5333479b3d640a2d292f42ef1da107f8ec4e83b

beaglebets reviewed (changes requested):
The overall theme here is consistent naming across libraries should be maintained, and that the default needs to remain sec_key so we don't break things for people when they deploy and accept a version bump.

danbernier committed:
Update json gem

I was getting local failures, but travis seems to have been fine? But I've been
running a lot of work and specs locally with this newer version of json, so it
seems safe.

danbernier committed:
Rename request-builder methods back to configure-foo-json to match JS library

danbernier committed:
Change option `:use_legacy_sec_key` to `:signature`

danbernier committed:
Drop specs for Utilities#request_security

The behavior is already well-covered by the other specs. This means we can drop
the request_time argument, too, even the ones that were copy-pasted for
consistency.

danbernier commented:
The build failed because I upgraded `json` on this branch, and that caused a conflict w/ the .travis.yml configuration. I've addressed that in an isolated PR: https://github.com/smileidentity/smile-identity-core-ruby/pull/15 Once that's merged in, I'll update here.

danbernier committed:
Merge branch 'signature' into signature-assist

Commit:
Merge branch 'signature' into signature-assist

Commit:
Point SDK at our domains, away from AWS

danbernier commented:
I've been testing this today. Here's how it's gone.

I got a new API key for "023" from dev:
```
partner1 % heroku run rails console --remote dev01-smileid
irb(main):003:0>  CompanyConfig.generate_api_key_v2 '023', ENV['SMILE_ENV'] # it was nil before
=> "abc123de-f45g-hijk-4321-1a2b3c4d5e6f"
```
In a local console on this branch, I ran some jobs and queries through, with that api_key:
```
dan@tower-of-babylon smile-identity-core-ruby % git branch
  master
  signature
* signature-assist
dan@tower-of-babylon smile-identity-core-ruby % bundle console
3.0.0 :001 > api_key = "abc123de-f45g-hijk-4321-1a2b3c4d5e6f"
 => "abc123de-f45g-hijk-4321-1a2b3c4d5e6f"
```

I tested submitting a job through IDAPI:
```
3.0.0 :002 > idapi = SmileIdentityCore::IDApi.new('023', api_key, 'https://devapi.smileidentity.com/v1') # NB that /v1
 => ...
3.0.0 :003 > pp JSON.parse(idapi.submit_job({"job_type": 5, "user_id": " ", "job_id": " "}, { "country": "Ghana", "id_type": "PASSPORT", "id_number": "G0000000",
 "phone_number": "1234567890", "partner_id": "023", "first_name": "some", "middle_name": " ", "last_name": "guy", "dob": "1990-01-31" }, signature: true)).reject { |k, v| v.to_s.size > 500 }.to_h
{"JSONVersion"=>"1.0.0",
 "SmileJobID"=>"0000014958",
 "PartnerParams"=>{"job_type"=>5, "user_id"=>" ", "job_id"=>" "},
 "ResultType"=>"ID Verification",
 "ResultText"=>"ID Number Validated",
 "ResultCode"=>"1012",
 "IsFinalResult"=>"true",
 "Actions"=>
  {"Verify_ID_Number"=>"Verified", "Return_Personal_Info"=>"Returned"},
 "Country"=>"Ghana",
 "IDType"=>"PASSPORT",
 "IDNumber"=>"G0000000",
 "ExpirationDate"=>"2021-07-13",
 "FullName"=>"Joe Doe Leo",
 "DOB"=>"2000-09-20",
 "PhoneNumber"=>"Not Available",
 "PhoneNumber2"=>"Not Available",
 "Document"=>"Not Available",
 "Gender"=>"MALE",
 "Address"=>"Not Available",
 "Source"=>"ID API",
 "signature"=>"2notarealsignaturenotarealsignature=",
 "timestamp"=>"2021-09-08T21:14:00.163Z"}
 => ...
```

I fetched a job's status (I got the job ID & user ID by looking in metabase):
```
3.0.0 :004 > pp SmileIdentityCore::Utilities.new('023', api_key, 'https://devapi.smileidentity.com/v1').get_job_status('57484385-6c4d-418f-b8c5-11845c4fa5f5', '9
b88d43d-7831-4a13-9e73-bb998aeaadb3', signature: true)
{"timestamp"=>"2021-09-08T21:21:34.752Z",
 "signature"=>"stuffthat/goes+in+here=",
 "job_complete"=>true,
 "job_success"=>true,
 "code"=>"2302",
 "result"=>
  {"Source"=>"SDK - Pixel XL",
   "Actions"=>
    {"Liveness_Check"=>"Passed",
     "Register_Selfie"=>"Approved",
     "Selfie_Provided"=>"Passed",
     "Verify_ID_Number"=>"Not Applicable",
     "Human_Review_Compare"=>"Not Applicable",
     "Return_Personal_Info"=>"Not Applicable",
     "Selfie_To_ID_Card_Compare"=>"Not Applicable",
     "Human_Review_Update_Selfie"=>"Not Applicable",
     "Human_Review_Liveness_Check"=>"Passed",
     "Selfie_To_ID_Authority_Compare"=>"Not Applicable",
     "Update_Registered_Selfie_On_File"=>"Not Applicable",
     "Selfie_To_Registered_Selfie_Compare"=>"Not Applicable"},
   "ResultCode"=>"1240",
   "ResultText"=>"Enroll User",
   "ResultType"=>"DIVA",
   "SmileJobID"=>"0000014755",
   "JSONVersion"=>"1.0.0",
   "IsFinalResult"=>"true",
   "PartnerParams"=>
    {"job_id"=>"9b88d43d-7831-4a13-9e73-bb998aeaadb3",
     "user_id"=>"57484385-6c4d-418f-b8c5-11845c4fa5f5",
     "job_type"=>"4"},
   "ConfidenceValue"=>"100",
   "IsMachineResult"=>"true"}}
 => ...
```

I'm still working on testing submitting a job through SmileIdentityCore::WebApi (that'll be for tomorrow).

danbernier commented:
I was able to successfully test WebApi#submit_job with a signature today. Here's a lightly-edited transcript of it:

```
dan@tower-of-babylon smile-identity-core-ruby % bundle console
3.0.0 :002 > empty_callback_url = nil
3.0.0 :003 > web_api = SmileIdentityCore::WebApi.new('023', empty_callback_url, same_api_key_from_yesterday, 'https://devapi.smileidentity.com/v1')
 => #<SmileIdentityCore::WebApi:0x00007fe3628681f0 @partner_id="023", @callback_url=nil, ...
3.0.0 :012 > partner_params = {
3.0.0 :013 >     job_type: 4,
3.0.0 :014 >     user_id: 'useriduseriduserid',
3.0.0 :015 >     job_id: 'jobidjobidjobid0',
3.0.0 :016 > }
 => {:job_type=>4, :user_id=>"useriduseriduserid", :job_id=>#<JobId:0x00007fe363673ce0 @count=0>}
3.0.0 :017 > images = [{image_type_id: 0, image: '/Users/dan/Desktop/Screen Shot 2021-06-01 at 3.36.19 PM.png'}]
 => [{:image_type_id=>0, :image=>"/Users/dan/Desktop/Screen Shot 2021-06-01 at 3.36.19 PM.png"}]
3.0.0 :018 > id_info = {}
3.0.0 :020 > pp web_api.submit_job(partner_params, images, id_info, signature: true, return_job_status: true); :dont_flood_repl
{"timestamp"=>"2021-09-09T16:50:36.321Z",
 "signature"=>"itsasignatureitsasignatureitsasignature=",
 "job_complete"=>true,
 "job_success"=>false,
 "code"=>"2302",
 "result"=>
  {"Source"=>"SDK",
   "Actions"=>
    {"Liveness_Check"=>"Not Applicable",
     "Register_Selfie"=>"Not Applicable",
     "Selfie_Provided"=>"Not Applicable",
     "Verify_ID_Number"=>"Not Verified",
     "Human_Review_Compare"=>"Not Applicable",
     "Return_Personal_Info"=>"Not Returned",
     "Selfie_To_ID_Card_Compare"=>"Not Applicable",
     "Human_Review_Update_Selfie"=>"Not Applicable",
     "Human_Review_Liveness_Check"=>"Not Applicable",
     "Selfie_To_ID_Authority_Compare"=>"Not Applicable",
     "Update_Registered_Selfie_On_File"=>"Not Applicable",
     "Selfie_To_Registered_Selfie_Compare"=>"Not Applicable"},
   "ResultCode"=>"0001",
   "ResultText"=>"Data Invalid",
   "ResultType"=>"SYSTEM",
   "SmileJobID"=>"0000014967",
   "JSONVersion"=>"1.0.0",
   "IsFinalResult"=>"true",
   "PartnerParams"=>
    {"job_id"=>"jobidjobidjobid0",
     "user_id"=>"useriduseriduserid",
     "job_type"=>"4"},
   "ConfidenceValue"=>"0",
   "IsMachineResult"=>"true"},
 "success"=>true,
 "smile_job_id"=>"0000014967"}
 => :dont_flood_repl
3.0.0 :021 >
```

At this point, I think this is ready to be merged and released.

danbernier commented:
@beaglebets I've addressed all the items you mentioned, and I've tested it thoroughly, afaict. Am I missing anything?
